### PR TITLE
Kubernetes Scheduler Support

### PR DESCRIPTION
**Pull request by jrcrawfo**
What?
Scheduler to support Kubernetes deployments

Why?
People using Kubernetes can deploy Heron

Notes:
Tests are coming up next.
Testing over the weekend around multiple containers per host to verify ports are working the way they should

### CODE REVIEW

**Comment by objmagic** [comment]
minor comment: copyright year should be 2017.

File: scripts/packages/BUILD [hunk 1]
```diff
@@ -251,6 +251,14 @@ pkg_tar(
 )
 
 pkg_tar(
+    name = "heron-client-conf-kubernetes",
+    package_dir = "conf/kubernetes",
+    files = [
+        "//heron/config/src/yaml:conf-kubernetes-yaml",
```
Comment (nlu90) [inline scripts/packages/BUILD]: stateful.yaml also needs to be added into the kubernetes conf diretory

File: heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesController.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+    } finally {
+      // Disconnect to release resources
+      conn.disconnect();
+    }
+
+
+  }
+
+  public boolean restartApp(int appId) {
```
Comment (jrcrawfo) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesController.java]: Done. In the controller

File: heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+    ObjectMapper mapper = new ObjectMapper();
+
+    // Align resources to maximal requested resource
+    PackingPlan updatedPackingPlan = packing.cloneWithHomogeneousScheduledResource();
+    SchedulerUtils.persistUpdatedPackingPlan(Runtime.topologyName(runtime),
+        updatedPackingPlan, Runtime.schedulerStateManagerAdaptor(runtime));
+
+    Resource containerResource = updatedPackingPlan.getContainers()
+        .iterator().next().getRequiredResource();
```
Comment (jrcrawfo) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java]: Done

File: scripts/packages/BUILD (hunk 1 above)
Comment (jrcrawfo) [inline scripts/packages/BUILD]: Merged master and took care of this

File: heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    ArrayNode commandsArray = mapper.createArrayNode();
+    for (int i = 0; i < command.length; i++) {
+      commandsArray.add(command[i]);
+    }
+    containerInfo.set(KubernetesConstants.COMMAND, commandsArray);
+
+    // Requested resource info
+    ObjectNode requestedResourceInfo = mapper.createObjectNode();
+    requestedResourceInfo.put(KubernetesConstants.MEMORY, containerResource.getRam().asMegabytes());
+    requestedResourceInfo.put(KubernetesConstants.CPU, containerResource.getCpu());
```
Comment (jrcrawfo) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java]: This is not yet supported in kubernetes

**Comment by jrcrawfo** [comment]
@nlu90 @kramasamy ready for a second review

File: heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java (hunk 4 above)
Comment (kramasamy) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java]: If this is not supported - when multiple containers/pods run in the same machine, how to do they divvy up the disk space - is there a formula?

Comment (jrcrawfo) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java]: I couldn’t find it immediately in the docs last night, but it seems like based on my reading there’s essentially no limit until the node runs out of space. They are planning to add this very soon though from what I read

Comment (kramasamy) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/KubernetesScheduler.java]: got it. We can add this once the feature is available -
https://github.com/twitter/heron/issues/1943

**Pull request merged**